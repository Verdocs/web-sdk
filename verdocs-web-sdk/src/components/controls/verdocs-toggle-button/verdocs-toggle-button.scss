@import '../../../theme.scss';

verdocs-toggle-button {
  font-family: $primary-font;

  button {
    border: none;
    cursor: pointer;
    align-items: center;
    display: inline-flex;
    justify-content: center;
    background: $verdocs-bg-1;
    color: $verdocs-dark-blue;

    svg {
      max-width: 100%;
      max-height: 100%;
      fill: $verdocs-dark-blue;
    }

    &.active {
      background: $verdocs-green;
      color: $verdocs-grey-3;

      svg {
        fill: $verdocs-grey-3;
      }
    }
  }

  &.size-normal button {
    width: 40px;
    height: 40px;
    padding: 6px;
    border-radius: 4px;
  }

  &.size-small button {
    width: 34px;
    height: 34px;
    padding: 4px;
    border-radius: 2px;
  }
}
