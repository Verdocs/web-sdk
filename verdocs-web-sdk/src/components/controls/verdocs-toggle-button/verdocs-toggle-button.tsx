import {Component, Prop, h, State, Event, EventEmitter, Host} from '@stencil/core';

/**
 * Displays a single button that can be toggled on or off by clicking it.
 */
@Component({
  tag: 'verdocs-toggle-button',
  styleUrl: 'verdocs-toggle-button.scss',
})
export class VerdocsToggleButton {
  @Prop() active = false;

  /**
   * If set, should be an SVG object. This will be rendered as the button's visible element. If icon is supplied, label is ignored.
   */
  @Prop() icon?: string | null = null;

  /**
   * If set, should be an SVG object. This will be rendered as the button's visible element. If icon is supplied, label is ignored.
   */
  @Prop() label?: string | null = null;

  /**
   * How large the button should be. Small buttons are intended for dialog boxes and other smaller scale UI regions.
   */
  @Prop() size?: 'small' | 'normal' = 'normal';

  /**
   * Event fired when the button is pressed.
   */
  @Event({composed: true}) toggle!: EventEmitter<{active: boolean}>;

  @State() _active = false;

  componentWillLoad() {
    this._active = this.active;
  }

  handleToggle() {
    const newState = !this._active;
    this._active = newState;
    this.toggle?.emit({active: newState});
  }

  render() {
    return (
      <Host class={`size-${this.size}`}>
        {this.icon ? (
          <button innerHTML={this.icon} class={{active: this._active}} onClick={() => this.handleToggle()} />
        ) : (
          <button class={{active: this._active}} onClick={() => this.handleToggle()}>
            {this.label}
          </button>
        )}
      </Host>
    );
  }
}
