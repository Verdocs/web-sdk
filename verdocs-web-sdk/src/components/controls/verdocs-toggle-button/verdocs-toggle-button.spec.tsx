import { newSpecPage } from '@stencil/core/testing';
import { VerdocsToggleButton } from './verdocs-toggle-button';

describe('verdocs-toggle-button', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsToggleButton],
      html: `<verdocs-toggle-button></verdocs-toggle-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders as active with icon and size', async () => {
    const icon = '<svg></svg>';
    const page = await newSpecPage({
      components: [VerdocsToggleButton],
      html: `<verdocs-toggle-button active="true" icon='${icon}' size="small"></verdocs-toggle-button>`,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.active).toBe(true);
    expect(page.rootInstance.icon).toBe(icon);
    expect(page.rootInstance.size).toBe('small');
  });

  it('emits toggle event on button click', async () => {
    const page = await newSpecPage({
      components: [VerdocsToggleButton],
      html: `<verdocs-toggle-button></verdocs-toggle-button>`,
    });
    const spy = jest.fn();
    page.root.addEventListener('toggle', spy);

    const button = page.root.querySelector('button');
    button?.dispatchEvent(new Event('click', { bubbles: true }));
    expect(spy).toHaveBeenCalled();
  });
});
