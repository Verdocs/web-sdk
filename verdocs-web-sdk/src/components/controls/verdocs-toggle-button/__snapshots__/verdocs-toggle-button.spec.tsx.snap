// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-toggle-button renders as active with icon and size 1`] = `
<verdocs-toggle-button active="true" class="size-small" icon="<svg></svg>" size="small">
  <button class="active">
    <svg></svg>
  </button>
</verdocs-toggle-button>
`;

exports[`verdocs-toggle-button renders with default props 1`] = `
<verdocs-toggle-button class="size-normal">
  <button></button>
</verdocs-toggle-button>
`;
